Modules/MUON/MCH/src/DecodingErrorsTask.cxx

```diff
@@ -0,0 +1,259 @@
+{
+}
+
+DecodingErrorsTask::~DecodingErrorsTask()
+{
+  if (mHistogramErrors) {
+    delete mHistogramErrors;
+  }
+}
+
+static void setAxisLabels(TH2F* hErrors)
+{
+  TAxis* ax = hErrors->GetXaxis();
+  for (int i = 1; i <= 10; i++) {
+    auto label = fmt::format("CH{}", i);
+    ax->SetBinLabel(i, label.c_str());
+  }
+  TAxis* ay = hErrors->GetYaxis();
+  //a->SetBit(TAxis::kLabelsHori);
+  ay->SetBinLabel(1, "Parity Error");
+  ay->SetBinLabel(2, "Hamming Error (correctable)");
+  ay->SetBinLabel(3, "Hamming Error (uncorrectable)");
+  ay->SetBinLabel(4, "Bad Cluster Size");
+  ay->SetBinLabel(5, "Bad Packet Type");
+  ay->SetBinLabel(6, "Bad HB Packet");
+  ay->SetBinLabel(7, "Bad Incomplete word");
+  ay->SetBinLabel(8, "Truncated Data");
+  ay->SetBinLabel(9, "Bad Elink ID");
+  ay->SetBinLabel(10, "Bad Link ID");
+  ay->SetBinLabel(11, "Unknown Link ID");
+  for (int i = 1; i <= 11; i++) {
+    ay->ChangeLabel(i, 45);
+  }
+}
+
+void DecodingErrorsTask::initialize(o2::framework::InitContext& /*ic*/)
+{
+  ILOG(Info) << "initialize DecodingErrorsTask" << ENDM; // QcInfoLogger is used. FairMQ logs will go to there as well.
+
+  mElec2Det = createElec2DetMapper<ElectronicMapperGenerated>();
+  mFee2Solar = createFeeLink2SolarMapper<ElectronicMapperGenerated>();
+
+  mHistogramErrors = new TH2F("QcMuonChambers_Errors", "Error codes vs. Chamber Number", 10, 1, 11, 10, 0, 10);
+  setAxisLabels(mHistogramErrors);
+  getObjectsManager()->startPublishing(mHistogramErrors);
+}
+
+void DecodingErrorsTask::startOfActivity(Activity& activity)
+{
+  ILOG(Info) << "startOfActivity : " << activity.mId << ENDM;
+  mHistogramErrors->Reset();
+}
+
+void DecodingErrorsTask::startOfCycle()
+{
+  ILOG(Info) << "startOfCycle" << ENDM;
+}
+
+void DecodingErrorsTask::decodeTF(framework::ProcessingContext& pc)
+{
+  // get the input buffer
+  auto& inputs = pc.inputs();
+  DPLRawParser parser(inputs, o2::framework::select("TF:MCH/RAWDATA"));
+
+  for (auto it = parser.begin(), end = parser.end(); it != end; ++it) {
+    auto const* raw = it.raw();
+    if (!raw) {
+      continue;
+    }
+    size_t payloadSize = it.size();
+
+    gsl::span<const std::byte> buffer(reinterpret_cast<const std::byte*>(raw), sizeof(RDH) + payloadSize);
+    decodeBuffer(buffer);
+  }
+}
+
+void DecodingErrorsTask::decodeReadout(const o2::framework::DataRef& input)
+{
+  // get the input buffer
+  if (input.spec->binding != "readout") {
+    return;
+  }
+
+  const auto* header = o2::header::get<header::DataHeader*>(input.header);
+  if (!header) {
+    return;
+  }
+
+  size_t payloadSize = header->payloadSize;
+  if (payloadSize == 0) {
+    return;
+  }
+
+  auto const* raw = input.payload;
+
+  gsl::span<const std::byte> buffer(reinterpret_cast<const std::byte*>(raw), payloadSize);
+  decodeBuffer(buffer);
+}
+
+void DecodingErrorsTask::decodeBuffer(gsl::span<const std::byte> buf)
+{
+  if (sDebug) {
+    std::cout << "\n\n============================\nStart of new buffer\n";
```

Comment:
If you plan to use the debug mode in the production setup (i.e. with InfoLogger), please use one of the logging facilities instead. Otherwise, it is fine with me.

Reply:
No, this is really just to debug the code during development. The output would be I think too verbose for the info logger.
However I am considering to convert some of the messages into infoLogger ones, to have some useful diagnostics in the production setup. If time allows I will look into that before this is merged.